### ISSUE TITLE
Bot does not reconnect after connection loss

### ISSUE THREAD

**Issue by HyperGH** [hikari-py/hikari#1011]
### Steps to reproduce

1. Start up a bot (tested with a GatewayBot with a single shard)
2. Trip over the ethernet cable
3. Watch logs

### Expected result

![image](https://user-images.githubusercontent.com/46067571/152839939-3706de7b-011b-4910-ae53-a730fbbbade7.png)
Exponentially backing off while trying to reconnect to the gateway. This screenshot was achieved via a hacky solution that has other issues, but demonstrates the desired behaviour.

### Actual result

![image](https://user-images.githubusercontent.com/46067571/152839657-2e696556-cac0-46a1-8721-c29676c1551a.png)
Apologies for the absolute garbage image quality, this is what not having an internet connection does to you.

TL;DR: The bot fails with an asyncio.TimeoutError after a single backoff.

### System info

```shell
hikari (2.0.0.dev106) [HEAD]
located at /home/hyper/.local/lib/python3.10/site-packages/hikari
CPython 3.10.2 GCC 11.1.0
Linux archlinux 5.16.5-zen1-1-zen #1 ZEN SMP PREEMPT Tue, 01 Feb 2022 21:42:42 +0000 x86_64
```

### Checklist

- [X] I have made sure to remove ANY sensitive information (bot token, passwords, credentials, personal details, etc.).
- [X] I have searched the issue tracker and have made sure it's not a duplicate. If it is a follow up of another issue, I have specified it.
